lib/src/widgets/board_editor.dart

```diff
@@ -0,0 +1,232 @@
+import 'package:chessground/src/board_geometry.dart';
+import 'package:flutter/widgets.dart';
+
+import 'piece.dart';
+import 'positioned_square.dart';
+import '../models.dart';
+import '../fen.dart';
+import '../board_editor_settings.dart';
+
+/// A chessboard widget whose pieces can be dragged around freely.
+///
+/// This widget can be used as the basis for a fully fledged board editor, similar to https://lichess.org/editor.
+class BoardEditor extends StatefulWidget with BoardGeometry {
+  const BoardEditor({
+    super.key,
+    required this.size,
+    required this.orientation,
+    required this.initialFen,
+    this.pieceToAddOnTap,
+    this.removePieceOnTap = false,
+    this.settings = const BoardEditorSettings(),
+    this.onPiecesChanged,
+  });
+
+  /// Visual size of the board.
+  @override
+  final double size;
+
+  /// Settings that control the appearance of the board editor.
+  final BoardEditorSettings settings;
+
+  /// Side by which the board is oriented.
+  @override
+  final Side orientation;
+
+  /// FEN string describing the initial position of the board editor.
+  final String initialFen;
+
+  /// If true, pieces will be removed when tapped. This takes precedence over [pieceToAddOnTap].
+  final bool removePieceOnTap;
```

Comment:
should we move `removePieceOnTap` and `pieceToAddOnTap` into a dedicated struct somehow? I'm not 100% happy with having two variables here, wish dart had `union`s

Reply:
Yes, I agree that this API is not optimal. 
I haven't tried it, but to me the best way to handle this is to externalise all the position update logic outside of the `BoardEditor` widget. The latter's only task should be to track user input.

Thinking of such an API:
```dart
class BoardEditor extends StatefulWidget {
  // FEN is the only source of truth for the position and it's handled outside of chessground (like the `Board` widget)
  // you need to implement logic in the `didUpdateWidget` to move the pieces (as in `Board`)
  // but we don't need anymore a `onPieceChange` callback
  final String fen;

  // Indicate the consumer that a square has been tapped;
  // the consumer should perform an action based on its own state: 
  // -put a piece here
  // -remove the piece
  // -do nothing if not in "tap" mode,
  // etc.
  final void Function(SquareId square) onTappedSquare;

  // Indicate the consumer a piece has been dropped on this square
  // If origin is not null it means a piece has been moved from the board
  // otherwise it means the piece comes from outside of the board
  final void Function(SquareId? origin, SquareId destination, Piece piece) onDroppedPiece;

  // indicate a piece on this square has been dragged outside (discarded)
  final void Function(SquareId square) onDiscardedPiece;
}
```

I did not take all the features into consideration so this is maybe not the definitive API, but you get the idea. And it solves the problems raised in your other comments.

With this API we don't support moving a piece by tap and tap, but it is better like that.

Reply:
With this API, the chessground consumer needs to maintain its own `Map<SquareId, Piece>`. So it needs 2 functions:
```dart
Pieces readFen(String fen);
String writeFen(Pieces pieces);
```

We already expose the first one; I think we need to expose the latter also, for convenience.

Reply:
It is maybe even possible to have a `StatelessWidget` here. FEN can be decoded in the `build` method since we don't need animations.

Reply:
Thanks for taking the time, really appreciate the feedback!

I pretty much went with the API you proposed, with the exception that I'm not using the `fen` as the single source of truth, but rather the `Pieces` (but the state is now outside the `BoardEditor`, as suggested).

This has the advantage that we do not need to constantly convert back and forth between FEN and `Pieces`. (If we passed the FEN instead, then the BoardEditor will parse it to `Pieces` in its `build()` method, but as soon as anything changes, the consumer makes the changes to its own `Piece` variable, and then call `writeFen` - Only for the widget calling `readFen` again.)

Eventually, the board editor implementation in the app will have to convert the `Pieces` to FEN, once the position has been confirmed by the user, but I think we can use `Board.fen` from dartchess for that, so we don't really need a `writeFen` in chessground for this.